akka-persistence/src/main/scala/akka/persistence/journal/WriteJournalBase.scala

```diff
@@ -4,21 +4,36 @@
 
 package akka.persistence.journal
 
-import akka.persistence.{ PersistentRepr, PersistentEnvelope }
+import akka.persistence.{ Persistence, PersistentRepr, PersistentEnvelope }
 import akka.actor.Actor
 import scala.collection.immutable
 
 private[akka] trait WriteJournalBase {
   this: Actor ⇒
 
+  private lazy val eventAdapters = Persistence(context.system).adaptersFor(context.self)
+
   protected def preparePersistentBatch(rb: immutable.Seq[PersistentEnvelope]): immutable.Seq[PersistentRepr] =
-    rb.filter(persistentPrepareWrite).asInstanceOf[immutable.Seq[PersistentRepr]] // filter instead of flatMap to avoid Some allocations
+    rb.collect {
+      case p: PersistentRepr ⇒
+        p.prepareWrite()
+        adaptToJournal(p)
+    } // collect instead of flatMap to avoid Some allocations
+
+  /** INTERNAL API */
+  private[akka] final def adaptFromJournal(repr: PersistentRepr): PersistentRepr = {
+    val adapter = eventAdapters.get(repr.payload.getClass)
```

Comment:
See https://github.com/ktoso/akka/commit/29d82b1875ab9a6639bb96bc7a25015f3d8c2714 for the "adapterId" implementation. Today we figured out during the meeting that even this is not enough, as it does not cover:

- initially no adapter
- now we need to add an adapter for some type
- that type could be "json", so it's hard to restrict to one